src/shared/components/oncoprint/DataUtils.spec.ts

```diff
@@ -0,0 +1,785 @@
+import {ClinicalAttribute, GeneMolecularData, Sample} from "../../api/generated/CBioPortalAPI";
+import {SpecialAttribute} from "../../cache/ClinicalDataCache";
+import {OncoprintClinicalAttribute} from "./ResultsViewOncoprint";
+import {MutationSpectrum} from "../../api/generated/CBioPortalAPIInternal";
+
+describe("DataUtils", ()=>{
+   describe("selectDisplayValue", ()=>{
+       it("returns undefined if no values", ()=>{
+           assert.equal(selectDisplayValue({}, {}), undefined);
+       });
+       it("returns the lone value if one value", ()=>{
+           assert.equal(selectDisplayValue({"a":0}, {"a":0}), "a");
+       });
+       it("returns the lowest priority value if two values", ()=>{
+           assert.equal(selectDisplayValue({"a":0, "b":0}, {"a":0, "b":1}), "a");
+           assert.equal(selectDisplayValue({"a":0, "b":0}, {"a":1, "b":0}), "b");
+       });
+       it("returns the lowest priority value if several values", ()=>{
+           assert.equal(selectDisplayValue({"a":0, "b":0, "c":5}, {"a":0, "b":1, "c":2}), "a");
+           assert.equal(selectDisplayValue({"a":20, "b":0, "c":10}, {"a":2, "b":1, "c":0}), "c");
+       });
+       it("returns the lowest priority, highest count value if two values w same priority", ()=>{
+           assert.equal(selectDisplayValue({"a":1, "b":0}, {"a":0, "b":0}), "a");
+           assert.equal(selectDisplayValue({"a":0, "b":1}, {"a":0, "b":0}), "b");
+       });
+       it("returns the lowest priority, highest count value if several values w same priority", ()=>{
+           assert.equal(selectDisplayValue({"a":1, "b":0, "c":5}, {"a":0, "b":0, "c":2}), "a");
+           assert.equal(selectDisplayValue({"a":20, "b":0, "c":10}, {"a":0, "b":1, "c":0}), "a");
+       });
+   });
+
+   describe("fillGeneticTrackDatum", ()=>{
+       it("fills a datum w no data correctly", ()=>{
+           assert.deepEqual(
+               fillGeneticTrackDatum({}, "gene", []),
+               {
+                   gene:"gene",
+                   data: [],
+                   disp_cna: undefined,
+                   disp_mrna: undefined,
+                   disp_prot: undefined,
+                   disp_mut: undefined
+               } as any);
+       });
+       it("fills a datum w one mutation data correctly", ()=>{
+           let data = [{
+               mutationType: "missense",
+               putativeDriver: true,
+               molecularProfileAlterationType: AlterationTypeConstants.MUTATION_EXTENDED
+           } as AnnotatedExtendedAlteration];
+           assert.deepEqual(
+               fillGeneticTrackDatum({}, "gene", data),
+               {
+                   gene:"gene",
+                   data: data,
+                   disp_cna: undefined,
+                   disp_mrna: undefined,
+                   disp_prot: undefined,
+                   disp_mut: "missense_rec"
+               } as any, "test1");
```

Comment:
Does `test1` correspond to the message in here http://chaijs.com/api/assert/? If so, it seems it's optional, so you don't need to send something like `test1`.

Reply:
I added it because when the tests were failing, the diff printing wasn't working properly so I couldn't tell which assertion within the `it` was failing.

Reply:
I thought yesterday we agreed on having only one case in each `it`, meaning whenever we call the tested method (`fillGeneticTrackDatum` in this case), we should have a separate `it` for it. I can see that it feels convenient to group some cases under one `it`, but with that logic one can create just one `it("should work correctly")` in every `describe`, so `it` would lose it's meaning. So as an example `it("should fill a datum for one putative missense mutation data correctly"` and `it("should fill a datum for one non-putative in_frame_del mutation data correctly"` looks better to me, and you wouldn't need the `test1` trick anymore.

Reply:
that seems unnecessary to me

Reply:
What is a broad case? Isn't every case is a separate case?

Reply:
if you read it you'll see, one case is "builds a datum correctly w a single mutation data", so it goes through a few ways for a single mutation data to happen.

Reply:
Yesterday you said you would split them up in the other PR, and it's the same situation here, so I thought we agreed on it. But not a big deal.

Reply:
I think its a good point in some cases but here unnecessary imo

Reply:
I think there are a couple decent reasons to group multiple assertions in same it.  One, I don't like an endless list of test results, so it's nice to be able to nest highly related assertions inside of one it (with use of assert message to distinguish failure, if dev isn't lazy ; ).  Two, if you need to run multiple assertions on the same returned item, you don't want to have to use closure to accomplish that.   An argument for one it one assert would be if the thing under test is stateful and dev should be forced to re-instantiate the thing for every test (using the beforeEach callback).   So, there are times when you definately should obey that rule.

Reply:
Good points 👍 Yeah as long as the assert message is meaningful, sounds good to me.